pkg/iter/cache.go

```diff
@@ -35,43 +36,52 @@ func (it *cachedIterator) Reset() {
 	it.curr = -1
 }
 
-func (it *cachedIterator) Next() bool {
-	if it.base != nil {
-		ok := it.base.Next()
-		// we're done with the base iterator.
-		if !ok {
-			it.closeErr = it.base.Close()
-			it.iterErr = it.base.Error()
-			it.base = nil
-			return false
-		}
-		// we're caching entries
-		it.cache = append(it.cache, entryWithLabels{entry: it.base.Entry(), labels: it.base.Labels()})
-		it.curr++
-		return true
+func (it *cachedIterator) Base() EntryIterator {
+	return it.base
+}
+
+func (it *cachedIterator) consumeNext() bool {
+	if it.base == nil {
+		return false
+	}
+	ok := it.base.Next()
+	// we're done with the base iterator.
+	if !ok {
+		it.closeErr = it.base.Close()
+		it.iterErr = it.base.Error()
+		it.base = nil
+		return false
 	}
-	// second pass
-	if len(it.cache) == 0 {
-		it.cache = nil
+	// we're caching entries
+	it.cache = append(it.cache, entryWithLabels{entry: it.base.Entry(), labels: it.base.Labels()})
+	it.curr++
+	return true
+}
+
+func (it *cachedIterator) Next() bool {
+	if len(it.cache) == 0 && it.base == nil {
 		return false
 	}
 	if it.curr+1 >= len(it.cache) {
+		if it.base != nil {
+			return it.consumeNext()
+		}
 		return false
 	}
 	it.curr++
-	return it.curr < len(it.cache)
+	return true
 }
 
 func (it *cachedIterator) Entry() logproto.Entry {
-	if len(it.cache) == 0 || it.curr < 0 {
+	if len(it.cache) == 0 || it.curr < 0 || it.curr >= len(it.cache) {
 		return logproto.Entry{}
```

Comment:
Hhmm, curious - why do we not return a pointer in this function (and `nil` in this case)?

Reply:
Would seem more memory-efficient to me to reference cached entries instead of copying

Reply:
This how the interface of the iterator is designed, I can't change that. But don't worry I did tried at some point and turns out it's not a big problem since there's a lot of optimization go does here under the hood.

Reply:
OK cool, thanks :+1: